### PR TITLE
[Modules] Connect Azure Firewall to a VirtualHub

### PR DESCRIPTION
**Pull request by fblix**
# Description

This PR introduces and fixes the functionality towards Azure Firewalls in combination with Secured Virtual Hubs.
A bit of background:
An Azure Firewall can be configured to either run in an VirtualNetwork setup (Sku: AzFw_Vnet) or in a Secured Virtual Hub context (Sku: AzFw_hub). A firewall can therefore be deployed without a virtual network and instead with a secured virtual hub.

---

It is also not possible to attach a Azure Firewall to a Virtual Hub but only in reverse order. If you allow the AzureFirewallId property in the virtual hub two things happen:

- The firewall is empty (has no policies attached): the deployment will succeed but not change anything.
- The firewall has policies attached: the deployment will fail with either:
  -  Standard SKU:  `{"code":"DeploymentFailed","message":"At least one resource deployment operation failed. Please list deployment operations for details. Please see https://aka.ms/DeployOperations for usage details.","details":[{"code":"AzureFirewallSkuCannotBeUpdated","message":"Update on SKU is not supported for Azure Firewall standard."}]}`
  - Premium SKU: `{"code":"DeploymentFailed","message":"At least one resource deployment operation failed. Please list deployment operations for details. Please see https://aka.ms/DeployOperations for usage details.","details":[{"code":"AzureFirewallSkuCannotBeUpdated","message":"Update on SKU is not supported for Azure Firewall premium (Sku name change is not allowed)."}]}`

Following the documentation [https://learn.microsoft.com/en-us/azure/firewall-manager/deployment-overview](here) reveals that there is a specific deployment order to be followed. Thus, we only enabled the connection between firewall and vhub to be established from the firewall side and NOT vice-versa to keep control of the deployment order

## Pipeline references

| Pipeline |
| --- |
| [Network: AzureFirewalls](https://github.com/Azure/ResourceModules/actions/workflows/ms.network.azurefirewalls.yml) |

# Type of Change

Please delete options that are not relevant.

- [x] Bugfix (non-breaking change which fixes an issue)

# Checklist

- [x] I'm sure there are no other open Pull Requests for the same update/change
- [x] My corresponding pipelines / checks run clean and green without any errors or warnings
- [x] My code follows the style guidelines of this project
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation (readme)
- [x] I did format my code

### CODE REVIEW

File: modules/Microsoft.Network/azureFirewalls/deploy.bicep [hunk 1]
```diff
@@ -177,6 +178,33 @@ var diagnosticsMetrics = [for metric in diagnosticMetricsToEnable: {
   }
 }]
 
+var firewallProperties = azureSkuName == 'AZFW_VNet' ? {
```
Comment (AlexanderSehr) [inline modules/Microsoft.Network/azureFirewalls/deploy.bicep:181]: Why this variable? I'd suggest to keep it close to the resource

Comment (AlexanderSehr) [reply modules/Microsoft.Network/azureFirewalls/deploy.bicep:181]: Other than that I'd good with merging the PR. Thanks @fblix & @JPEasier

Comment (fblix) [reply modules/Microsoft.Network/azureFirewalls/deploy.bicep]: good point. I guess it was a leftover from testing. removed it.

**Review by AlexanderSehr — APPROVED** [review]

File: modules/Microsoft.Network/azureFirewalls/deploy.bicep [hunk 2]
```diff
@@ -243,6 +244,18 @@ resource azureFirewall 'Microsoft.Network/azureFirewalls@2021-08-01' = {
     applicationRuleCollections: applicationRuleCollections
     natRuleCollections: natRuleCollections
     networkRuleCollections: networkRuleCollections
+  } : {
+    firewallPolicy: empty(firewallPolicyId) ? null : {
```
Comment (AlexanderSehr) [inline modules/Microsoft.Network/azureFirewalls/deploy.bicep:248]: Flip the condition please. It's more intuitive to read if the value comes first, then the default

Comment (AlexanderSehr) [reply modules/Microsoft.Network/azureFirewalls/deploy.bicep:248]: Same for the other conditions. I see a few of those where `null` is mentioned first

File: modules/Microsoft.Network/azureFirewalls/.test/hub/dependencies.bicep [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+  location: location
+  properties: {
+    threatIntelMode: 'Alert'
+  }
+}
+
+@description('The resource ID of the created Virtual Hub.')
+output virtualHubResourceId string = virtualHub.id
+
+@description('The resource ID of the created Firewall Policie.')
```
Comment (AlexanderSehr) [inline modules/Microsoft.Network/azureFirewalls/.test/hub/dependencies.bicep:42]:
```suggestion
@description('The resource ID of the created Firewall Policy.')
```
Nice try 😉

Comment (AlexanderSehr) [reply modules/Microsoft.Network/azureFirewalls/.test/hub/dependencies.bicep:42]: Addressed in #2547

File: modules/Microsoft.Network/azureFirewalls/deploy.bicep (hunk 2 above)
Comment (AlexanderSehr) [reply modules/Microsoft.Network/azureFirewalls/deploy.bicep:248]: Addressed in #2547

**Pull request merged**